db-tests/src/test/java/gov/nasa/jpl/aerie/database/ExternalEventTests.java

```diff
@@ -0,0 +1,1611 @@
+          assertTrue(res.next());
+          assertEquals("a", res.getString("event_key"));
+          assertEquals("00:20:00", res.getString("duration"));
+          assertEquals("C", res.getString("source_key"));
+          assertEquals("{[\"2024-01-01 00:00:00+00\",\"2024-01-01 01:00:00+00\")}", res.getString("source_range"));
+          assertFalse(res.next());
+        });
+      }
+    }
+  }
+
+
+  /*
+    ---------- CONSTRAINT TESTS ----------
+   */
+  /**
+   * The focus of this class is to provide a final set of tests to test primary and significant constraints to prove
+   *    that they work and how they work. These constraints are baked into the schema, so extensive tests of foreign or
+   *    unique keys are not excluded except for common use cases/easily made mistakes, as PSQL guarantees functionality
+   *    for us.
+   * We seek to demonstrate in this class, and verify as a sanity check.
+   */
+  @Nested
+  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+  class ConstraintTests {
+
+    /**
+     * Generally, no two sources in a derivation group, regardless of a lack of overlap, can have the same valid_at. We
+     *    demonstrate this with sparse sources. In this test, we don't even require events to demonstrate this behavior.
+     */
+    @Test
+    void sameValid_at() {
+      final var statement = getStatement();
+
+      // construct the sources
+      ExternalSource a = new ExternalSource(
+          "A",
+          st,
+          dg,
+          "2024-01-01T00:00:00Z",
+          "2024-01-01T03:00:00Z",
+          "2024-01-01T04:00:00Z",
+          ca,
+          mt);
+      ExternalSource b = new ExternalSource(
+          "B",
+          st,
+          dg,
+          "2024-01-01T00:00:00Z",
+          "2024-01-01T01:00:00Z",
+          "2024-01-01T02:00:00Z",
+          ca,
+          mt);
+
+      // create types and first source
+      assertDoesNotThrow(() -> {
+        statement.get().executeUpdate(
+            //language-sql
+            """
+            INSERT INTO merlin.external_source_type VALUES ('%s')
+            """.formatted(st)
+        );
+
+        statement.get().executeUpdate(
+            //language-sql
+            """
+            INSERT INTO merlin.derivation_group VALUES ('%s', '%s')
+            """.formatted(dg, st)
+        );
+
+        statement.get().executeUpdate(
+            // language-sql
+            """
+            INSERT INTO merlin.external_source VALUES ('%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s')
+            """.formatted(
+            a.key(),
+            a.source_type_name(),
+            a.derivation_group_name(),
+            a.valid_at(),
+            a.start_time(),
+            a.end_time(),
+            a.created_at(),
+            a.metadata()
+          )
+        );
+      });
+
+      // second source should fail
+      assertThrowsExactly(PSQLException.class, () -> {
+        statement.get().executeUpdate(
+            // language-sql
```

Comment:
Typo:
```suggestion
            // language=sql
```

This typo is widespread throughout this file. Please update all occurrences.